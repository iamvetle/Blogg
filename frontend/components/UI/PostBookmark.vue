<template>
    <div>
        <BaseIconSaveArticleSaved v-if="checkIfPostIsSaved(post)" @click="unsave(post)" />

        <BaseIconSaveArticleUnSaved v-else @mouseover="color = 'fill-primary'" @mouseleave="color = 'fill-black'"
            @click="save(post)" :fill-color="color" />
    </div>
</template>

<script setup lang="ts">
/**
 * This component handles the functionality for saving and unsaving articles. 
 * It shows different icons based on the saved state of the article.
 */

// Props definition
// @prop post - Numeric ID of the article.
defineProps<{
    post: number;
}>();

// Reactive variable for icon color
const color = ref("fill-black");

// Importing the loggedInUserStore from the store directory
import { useLoggedInUserStore } from '~/store/loggedInUserStore';

// Using the loggedInUserStore for managing the saved posts
const loggedInUserStore = useLoggedInUserStore();

/**
 * Removes the selected post from the saved posts list.
 * @param post - Numeric ID of the post to unsave.
 */
const unsave = async (post: number) => {
    const index = loggedInUserStore.idArrayOfSavedPosts.findIndex((id) => id === post);
    loggedInUserStore.idArrayOfSavedPosts.splice(index, 1);
    await getSaveOrUnsavePost(post);
}

/**
 * Adds the selected post to the saved posts list.
 * @param post - Numeric ID of the post to save.
 */
const save = async (post: number) => {
    await getSaveOrUnsavePost(post);
}

// Note: The function getSaveOrUnsavePost is used but not defined or imported in this snippet.
// Ensure it is properly defined and imported for the component to function correctly.
</script>

<style scoped></style>