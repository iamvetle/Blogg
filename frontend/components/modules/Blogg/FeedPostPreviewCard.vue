<template>
	<div class="article grid-cols-12 grid gap-8 mb-16">

		<div id="left-col-article" class="col-start-1 col-end-9">
			
			<span class="flex items-center w-full mb-2">
				<span class="user-picture flex items-center">
					<slot name="profile_picture"></slot>
				</span>
				<span class="flex items-center justify-between w-full">

					<span class="flex items-center full_name">
						<slot name="author"></slot>

					</span>

					<slot name="date_published"></slot>

				</span>
			</span>
			<div class="mb-4 w-full">

				<slot name="title"></slot>


				<slot name="content"></slot>

			</div>

			<span class="flex items-center justify-between">
				<span class="flex items-center space-x-2">
					<slot name="lesmer"></slot>
				
				<span class="flex items-center">
					<slot name="tags"></slot>
				</span>
				</span>

				<span class="flex items-center">
					<span id="comments-count" class="me-2">
						<slot name="amount-of-comments"></slot>
					</span>

					<span id="save-article" class="me-2">
						<slot name="save-article-icon"></slot>
					</span>

					<span id="article-options" class="">

						<slot name="more-options-icon"></slot>

					</span>

				</span>
			</span>
		</div>

		<div id="right-col-article" class="col-span-4 flex items-center justify-center">
			<slot name="article_image">This is were the article image is.</slot>
		</div>
	</div>
</template>

<script setup lang="ts">

// temp solution


</script>

<style scoped></style>