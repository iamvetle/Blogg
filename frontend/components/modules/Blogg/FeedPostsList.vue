<template>
	<div
		v-if="postStore.posts"
		id="site-wrapper"
		class="bg-background text-onBackground"
	>
		<div v-if="postStore.posts.results" id="content">
			<div class="article" v-for="post in postStore.posts.results">
				<feed-post-preview-card :key="post.id">
					<template #profile_picture>
						<img
							v-show="true"
							:src="account_picture"
							alt="Bruker profilbilde"
							class="w-6 h-auto me-2"
						/>
					</template>

					<template #author v-if="post.author">
						<span
							@click="
								redirect_to_author_page(post.author.username)
							"
							class="cursor-pointer break-words"
						>
							<p
								class="font-bold"
								v-text="author_full_name(post.author)"
							></p>
						</span>
					</template>

					<template #date_published v-if="post.date_published">
						<span>
							<p
								class="font-light"
								v-text="post.date_published"
							></p>
						</span>
					</template>

					<template #title v-if="post.title">
						<span
							@click="redirect_to_post_page(post.id)"
							class="cursor-pointer break-words"
						>
							<h3
								class="text-[28px] mb-2"
								v-text="post.title"
							></h3>
						</span>
					</template>

					<template #content v-if="post.content_snippet">
						<div class="break-words">
							<p
								class="mb-2"
								v-text="toPlainText(post.content_snippet)"
							></p>
						</div>
					</template>

					<template #lesmer v-if="post.id">
						<span
							class="cursor-pointer text text-primary hover:text-primaryFixed"
							@click="redirect_to_post_page(post.id)"
						>
							Les mer
						</span>
					</template>

					<!-- Needs to be authenticated to be able to save posts-->
					<template
						#tags
						v-if="post.tags && authStore.isAuthenticated"
					>
						<span class="me-1" v-for="tag in post.tags">
							<BaseTag :key="post.id" :text="tag.name" />
						</span>
					</template>

					<template
						#amount-of-comments
						v-if="post.num_of_comments !== null"
					>
						<span>{{ post.num_of_comments }} comments</span>
					</template>

					<template
						#save-article-icon
						v-if="
							post.id &&
							checkIfLoggedInUser(post.author.username) ==
								false &&
							authStore.isAuthenticated
						"
					>
						<PostBookmark
							:post="post.id"
							data-test="post_bookmark"
						/>
					</template>

					<template #more-options-icon v-if="post.id">
						<BaseIconMoreOptions
							widthProp="24"
							heightProp="24"
							:colorProp="color"
							@mouseover="color = 'fill-primary'"
							@mouseleave="color = 'fill-black'"
						/>
					</template>

					<template #article_image v-if="post.id">
						<img
							:src="post_image"
							alt="Bilde til artikkel"
							class="w-full h-auto"
						/>
					</template>
				</feed-post-preview-card>

				<hr class="mb-16" />
			</div>
		</div>
	</div>
</template>

<script setup lang="ts">
import account_picture from "~/assets/account-pin-circle-line.svg";

const post_image = ref("https://picsum.photos/500/300");

const postStore = usePostStore();
const authStore = useAuthStore();

const color = ref("fill-black");

/**
 * The function takes the html input and returns only the raw text
 * @returns - the raw text of the html
 */
const toPlainText = (htmlContent: string) => {
	const div = document.createElement("div");
	div.innerHTML = htmlContent;
	const rawText = div.textContent || div.innerText;
	return rawText;
};

/**
 * Redirects the web client to the profile page of the author
 * @param username
 */
const redirect_to_author_page = (username: string) => {
	return navigateTo(`/user/${username}`);
};

/**
 * Redirects the web client to the page of the post
 * @param post
 */
const redirect_to_post_page = (postId: number) => {
	return navigateTo(`/post/${postId}`);
};

/**
 *
 * @param author - the 'author' part of the relevant 'post'
 */
const author_full_name = (author: AuthorType) => {
	// console.log(author) // print to self

	const full_name = `${author.first_name} ${author.last_name}`;
	return full_name.trim() == "" ? author.username : full_name;
};

const paginationStore = usePaginationStore();

paginationStore.activeFetchURL = urls.api.posts.feed;

await getPostMultipleSnippet(paginationStore.activeFetchURL);

</script>

<style scoped></style>
