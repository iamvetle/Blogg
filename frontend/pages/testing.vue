<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>