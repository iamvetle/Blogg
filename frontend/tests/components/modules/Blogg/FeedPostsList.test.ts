
import { VueWrapper, mount } from '@vue/test-utils'
import FeedPostsList from '~/components/modules/Blogg/FeedPostsList.vue'
import { createTestingPinia } from '@pinia/testing'
import PostBookmark from '~/components/UI/PostBookmark.vue'
import FeedPostPreviewCard from '~/components/modules/Blogg/FeedPostPreviewCard.vue'

describe("list articles testing", () => {
    let wrapper: VueWrapper
    let generalStore: any;
    let postStore:any;
    let pinia: any;

    beforeEach(() => { // The beforeEach block runs before each test
        // Create a new testing Pinia instance
        pinia = createTestingPinia()

        vi.stubGlobal("checkIfLoggedInUser", () => {
            return false
        })

        // Initialize the generalStore with mock data
        generalStore = useGeneralStore(pinia)
        postStore = usePostStore(pinia)

        generalStore.isAuthenticated = true
        postStore.posts = {
            "count": 31,
            "next": "http://127.0.0.1:8888/api/feed/?page=3",
            "previous": "http://127.0.0.1:8888/api/feed/",
            "current_page": 2,
            "results": [
                {
                    "id": 17,
                    "title": "This is a title that is made by bob",
                    "author": {
                        "username": "bob",
                        "first_name": "Bob",
                        "last_name": "Smith",
                    },
                    "content_snippet": "Lorem ipsum...",
                    "date_published": "08-12-2021",
                    "num_of_comments":526,
                    "images":[
                        {
                            image:"~/src/something/image.jpeg"
                        },
                        {
                            image:"~/src/fruit/life.png"
                        }
                    ]
                },
                {
                    "id": 19,
                    "title": "This is a title James",
                    "author": {
                        "username": "jammy",
                        "first_name": "James",
                        "last_name": "Brown",
                    },
                    "content_snippet": "Test content content yes",
                    "date_published": "18-10-2022",
                    "num_of_comments":98,
                    "images":[
                        {
                            image:"~/src/africa/elefant.jpeg"
                        },
                        {
                            image:"~/src/vegetables/pingpong.png"
                        }
                    ]
                }
            ]
        }

        // Mount the component
        wrapper = mount(FeedPostsList, {
            global: {
                plugins: [pinia],
                components: {
                    FeedPostPreviewCard,
                    PostBookmark
                },
                stubs: {
                    PostBookmark:true,
                    BaseTag:true,
                    BaseIconMoreOptions:true,

                }
            },
            props: {
                width: "23",
                height:"51",
                fillColor:"blue",
                class:'',
            },
        })
    })

    test("exists", () => {
        expect(wrapper.exists()).toBe(true)
    })

    test("the articles/posts are rendering", () => {
        expect(wrapper.text()).toContain("This is a title that is made by bob")
        expect(wrapper.text()).toContain("bob")
        expect(wrapper.text()).toContain("Bob")
        expect(wrapper.text()).toContain("Smith")
        expect(wrapper.text()).toContain("Lorem ipsum...")
        expect(wrapper.text()).toContain("08-12-2021")
    })
    /**
     * ! try to mock the pagination store and such so that I can check more proper behavior through computed properties
     */
    // test("renders the article bookmark icon component", () => {
    //     const bookMarkComp = wrapper.find("[data-test='post_bookmark']") 

    //     expect(bookMarkComp.exists()).toBe(true)

    // })

    test("Should display the number of comments", () => {
        expect(wrapper.text()).toContain("526")
    })
    test('Should match snapshot', () => {
        expect(FeedPostsList).toMatchSnapshot()
    })
    
    
})
