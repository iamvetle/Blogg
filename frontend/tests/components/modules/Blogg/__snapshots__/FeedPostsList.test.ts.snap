// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list articles testing > Should match snapshot 1`] = `
{
  "__file": "C:/Users/metle/Desktop/Blogg/frontend/components/modules/Blogg/FeedPostsList.vue",
  "__name": "FeedPostsList",
  "render": [Function],
  "setup": [Function],
}
`;
